lib/puppet/provider/hocon_setting/ruby.rb

```diff
@@ -11,21 +11,36 @@ def self.namevar(section_name, setting)
   end
 
   def exists?
-    ret_value = false
+    unless conf_file.has_value?(setting)
+      return false
+    end
 
-    if conf_file.has_value?(setting)
-      if resource[:type] == 'array_element'
-        Array(@resource[:value]).each do |v|
-          if value.flatten.include?(v)
-            return true
-          end
+    type = @resource[:type]
+    conf_value = conf_object.get_value(setting).value
+
+    if type == 'array'
+      unless conf_value.is_a?(Array)
+        return false
+      end
+    end
+
+    if type == nil &&
+       conf_value.is_a?(Array) &&
+       conf_value.length == 1
+
+      return false
+    end
+
+    if type == 'array_element'
+      Array(@resource[:value]).each do |v|
```

Comment:
I think this can be `return Array(@resource[:value]).any? {|v| value.flatten.include?(v) }`, but I'm not sure if that makes it more readable or not.